File: moxy/src/main/java/com/omegar/mvp/viewstate/ViewCommand.java
package com.omegar.mvp.viewstate;

import com.omegar.mvp.MoxyReflector;
import com.omegar.mvp.MvpView;
import com.omegar.mvp.viewstate.strategy.AddToEndSingleStrategy;
import com.omegar.mvp.viewstate.strategy.SkipStrategy;
import com.omegar.mvp.viewstate.strategy.StateStrategy;

import java.io.Serializable;

/**
 * Date: 16-Dec-15
 * Time: 16:59
 *
 * @author Alexander Blinov
 */
public abstract class ViewCommand<View extends MvpView> {
	private final String mTag;
	private final StateStrategy mStateStrategy;

	protected ViewCommand(String tag, StateStrategy stateStrategy) {
		mTag = tag;
		mStateStrategy = stateStrategy;
	}

	// default constructor for serializable
	protected ViewCommand() {
		this("", AddToEndSingleStrategy.INSTANCE);
	}

	public abstract void apply(View view);

	public String getTag() {
		return mTag;
	}

	public StateStrategy getStateStrategy() {
		return mStateStrategy;
	}

	@Override
	public String toString() {
		return getClass().getSimpleName();
	}
}

File: moxy/src/main/java/com/omegar/mvp/viewstate/ViewCommand.kt
package com.omegar.mvp.viewstate

import com.omegar.mvp.MvpView
import com.omegar.mvp.viewstate.strategy.AddToEndSingleStrategy
import com.omegar.mvp.viewstate.strategy.StateStrategy

/**
 * Date: 16-Dec-15
 * Time: 16:59
 *
 * @author Alexander Blinov
 */
abstract class ViewCommand<View : MvpView> protected constructor(
        val tag: String = "",
        private val stateStrategy: StateStrategy = AddToEndSingleStrategy
) {

    fun beforeApply(commands: MutableList<ViewCommand<View>>) = stateStrategy.beforeApply(commands, this)

    abstract fun apply(mvpView: View)

    fun afterApply(commands: MutableList<ViewCommand<View>>) = stateStrategy.afterApply(commands, this)

    override fun toString(): String = javaClass.simpleName

    protected fun buildString(name: String, vararg args: Any?): String {
        return (0 until args.size / 2)
                .joinToString(prefix = "$name{", separator = ", ", postfix = "}") { i ->
                    val key = args[i * 2]
                    val value = args[i * 2 + 1]
                    val mark = "'".takeIf { value is String } ?: ""
                    "$key=$mark$value$mark"
                }
    }
}